File: spark1x-extension/src/main/scala/com/webex/dap/spark/core/FlumeIssueFix.scala
package com.webex.dap.spark.core

import com.google.gson.JsonParser
import org.apache.hadoop.fs.{FileSystem, Path}
import org.apache.hadoop.io.compress.LzoCodec
import org.apache.spark.{SparkConf, SparkContext}
import org.slf4j.LoggerFactory

import scala.collection.mutable
import scala.util.Random

/**
  * Created by harry on 3/5/18.
  */
object FlumeIssueFix {
  val logger = LoggerFactory.getLogger(FlumeIssueFix.getClass)
  val jsonParser = new JsonParser()

  val random = new Random()

  def main(args: Array[String]) {
    if (args.length < 4) {
      System.err.println("Usage: FlumeIssueFix <in-directoy> <day> <out-directoy> <size> <type#featurename#metricname>")
      System.exit(1)
    }

    val sparkConf = new SparkConf().setAppName("FlumeIssueFix")
    val sc = new SparkContext(sparkConf)

    //    val dfs = FileSystem.get(sc.hadoopConfiguration)

    //    val output = new Path(args(2));
    //    if (dfs.exists(output)){
    //      dfs.delete(output,true);
    //    }

    val lines = sc.textFile(args(0))
    var metrics = lines.map { line =>
      val keyMap = new mutable.HashMap[String, String]();
      try {
        val json = jsonParser.parse(line).getAsJsonObject;
        if (json.has("type") && !json.get("type").isJsonNull) {
          keyMap.put("type", json.get("type").getAsString().toLowerCase());
        }

        if (json.has("message") && !json.get("message").isJsonNull) {
          if (json.get("message").isJsonObject) {
            val message = json.get("message").getAsJsonObject
            if (message.has("featureName") && message.get("featureName").isJsonPrimitive) {
              keyMap.put("featureName", message.get("featureName").getAsString().toLowerCase());
            }
            if (message.has("metricName") && message.get("metricName").isJsonPrimitive) {
              keyMap.put("metricName", message.get("metricName").getAsString().toLowerCase());
            }
          }
        }
      } catch {
        case e =>
      }

      (keyMap, line)
    }

    val filter_ = args(4).split("#",3);

    if (filter_(0) != null && filter_(0) != "") {
      metrics = metrics.filter(_._1.getOrElse("type", "others").equals(filter_(0)));
    }

    if (filter_(1) != null && filter_(1) != "") {
      metrics = metrics.filter(_._1.getOrElse("featureName", "others").equals(filter_(1)));
    }

    if (filter_(2) != null && filter_(2) != "") {
      metrics = metrics.filter(_._1.getOrElse("metricName", "others").equals(filter_(2)));
    }

    val records = metrics.map(record => {
      (new StringBuilder().append(record._1.getOrElse("type", "others")).append("/").append(record._1.getOrElse("featureName", "others")).append("-").append(record._1.getOrElse("metricName", "others")).append("/day=").append(args(1)).append("##").append(random.nextInt(args(3).toInt)).toString(), record._2)
    }).reduceByKey((msg_1, msg_2) => msg_1 + "\n" + msg_2)

    records.saveAsHadoopFile(args(2), classOf[String], classOf[String], classOf[FlumeIssueRDDMultipleTextOutputFormat[_, _]], classOf[LzoCodec]);

    sc.stop()
  }
}

File: data/src/main/java/com/webex/dap/data/flume_/FlumeIssueFix.java
package com.webex.dap.data.flume_;

import com.google.gson.JsonObject;
import com.google.gson.JsonParser;

import java.util.HashMap;
import java.util.Map;

/**
 * Created by harry on 9/21/18.
 */
public class FlumeIssueFix {
    static JsonParser jsonParser = new JsonParser();

    public static void main(String[] args) {

        String line = "{\"type\": \"sdf\",\"message\": {\"metricName\"\": \"sdf\",\"featureName\": \"xfd\"}}";
        parse(line);
        line = "{\"type\": \"sdf\",\"message\": {\"metricName\": \"sdf\",\"featureName\": \"xfd\"}}";
        parse(line);
        line = "{\"type\": \"sdf\",\"message\": {\"metricName\": \"sdf\",\"featureName\": \"\"}}";
        parse(line);
        line = "{\"type\": \"sdf\",\"message\": {\"metricName\": \"sdf\"}}";
        parse(line);
        line = "{\"type\": \"sdf\",\"message\": {\"metricName\": \"\",\"featureName\": \"xfd\"}}";
        parse(line);
        line = "{\"type\": \"sdf\",\"message\": {\"featureName\": \"xfd\"}}";
        parse(line);
        line = "{\"type\": \"\",\"message\": {\"metricName\": \"sdf\",\"featureName\": \"xfd\"}}";
        parse(line);
        line = "{\"message\": {\"metricName\": \"sdf\",\"featureName\": \"xfd\"}}";
        parse(line);

        line = "{\"type\": \"sdf\",\"message\": {\"metricName\": \"\",\"featureName\": \"\"}}";
        parse(line);

        line = "{\"type\": \"sdf\",\"message\": {}}";
        parse(line);

        line = "{\"type\": \"\",\"message\": {\"metricName\": \"\",\"featureName\": \"xfd\"}}";
        parse(line);

        line = "{\"message\": {\"featureName\": \"xfd\"}}";
        parse(line);

        line = "{\"type\": \"\",\"message\": {\"metricName\": \"\",\"featureName\": \"\"}}";
        parse(line);

        line = "{\"message\": {\"metricName\": \"\"}}";
        parse(line);

    }

    public static void parse(String line) {
        System.out.println();
        Map<String, String> keyMap = new HashMap<String, String>();
        try {
            JsonObject json = jsonParser.parse(line).getAsJsonObject();
            if (json.has("type")) {
                keyMap.put("type", json.get("type").getAsString().toLowerCase());
            }

            if (json.has("message") && !json.get("message").isJsonNull()) {
                JsonObject message = json.get("message").getAsJsonObject();

                if (message.has("featureName")) {
                    keyMap.put("featureName", message.get("featureName").getAsString().toLowerCase());
                }

                if (message.has("metricName")) {
                    keyMap.put("metricName", message.get("metricName").getAsString().toLowerCase());
                }
            }
        } catch (Exception e) {
//            e.printStackTrace();
        } finally {
            if (!keyMap.containsKey("type")) {
                keyMap.put("type", "others");
            }
            if (!keyMap.containsKey("metricName")) {
                keyMap.put("metricName", "others");
            }
            if (!keyMap.containsKey("featureName")) {
                keyMap.put("featureName", "others");
            }
        }

        for (Map.Entry<String, String> xx : keyMap.entrySet()) {
            System.out.println(xx.getKey() + " -> " + xx.getValue());
        }
        System.out.println();
    }
}
